pkg/events/queue/queue_mem_list.go

```diff
@@ -0,0 +1,118 @@
+package queue
+
+import (
+	"container/list"
+	"fmt"
+	"sync"
+
+	"github.com/aquasecurity/tracee/types/trace"
+)
+
+type eventQueueMem struct {
+	mutex                *sync.Mutex
+	cond                 *sync.Cond
+	cache                *list.List
+	maxAmountOfEvents    int // max number of cached events possible
+	eventsCacheMemSizeMB int
+	verbose              string
+}
+
+func NewEventQueueMem(queueSizeMb int) CacheConfig {
+	q := &eventQueueMem{
+		eventsCacheMemSizeMB: queueSizeMb,
+	}
+	q.setup()
+	q.verbose = fmt.Sprintf("In-Memory Event Queue (Size = %d MB)", q.eventsCacheMemSizeMB)
+	return q
+}
+
+func (q *eventQueueMem) String() string {
+	return q.verbose
+}
+
+func (q *eventQueueMem) setup() {
+	q.mutex = new(sync.Mutex)
+	q.cond = sync.NewCond(q.mutex)
+
+	// set queue size and init queue
+	q.maxAmountOfEvents = q.getQueueSizeInEvents()
+	q.cache = new(list.List)
+}
+
+// Enqueue pushes an event into the queue (may block until queue is available)
+func (q *eventQueueMem) Enqueue(evt *trace.Event) {
+	q.cond.L.Lock()
+	// enqueue waits for de-queuing if cache is full (using >= instead of == to be in the safe side...)
+	for q.cache.Len() >= q.maxAmountOfEvents {
+		q.cond.Wait()
+	}
+
+	q.cache.PushBack(*evt)
+	q.cond.L.Unlock()
+	q.cond.Signal() // unblock dequeue if needed
+
+	evt = nil
+}
+
+// Dequeue pops an event from the queue
+func (q *eventQueueMem) Dequeue() *trace.Event {
+	q.cond.L.Lock()
+
+	// dequeue waits for en-queueing if cache is empty
+	for q.cache.Len() == 0 {
+		q.cond.Wait()
+	}
+
+	e := q.cache.Front()
+	event, ok := e.Value.(trace.Event)
+	if !ok {
+		q.cond.L.Unlock()
+		return nil
+	}
+	q.cache.Remove(e)
+	q.cond.L.Unlock()
+	q.cond.Signal() // unblock enqueue if needed
+
+	return &event
+}
+
+// getQueueSizeInEvents returns size of the fifo queue, in # of events, based on
+// the host size
+func (q *eventQueueMem) getQueueSizeInEvents() int {
+	// eventSize is the memory footprint per event, in bytes. This is NOT the
+	// size of a single event, but the overall impact in memory consumption to
+	// each cached event (defined by experimentation)
+	eventSize := 1024
+
+	KBtoB := func(amountInKB int) int {
+		return amountInKB * 1024
+	}
+	MBtoKB := func(amountInMB int) int {
+		return amountInMB * 1024
+	}
+	GBtoMB := func(amountInGB int) int {
+		return amountInGB * 1024
+	}
+	AmountOfEvents := func(amountInMB int) int {
+		return MBtoKB(KBtoB(amountInMB)) / eventSize
+	}
+
+	// EventsCacheMemSize was provided, return exact amount of events for it
+	if q.eventsCacheMemSizeMB > 0 {
+		return AmountOfEvents(q.eventsCacheMemSizeMB)
+	}
+
+	switch {
```

Comment:
we get here only if `q.eventsCacheMemSizeMB < = 0` and the only relevant switch case will always be the first one (`case q.eventsCacheMemSizeMB <= GBtoMB(1)`) . what am I missing? again, asking it here as this was part of your commits @rafaeldtinoco and maybe I am loosing something important in the logics.

Reply:
If you provide a cache size ("mem-cache-size" cmdline option), then you know what you are doing... q.eventsCacheMemSizeMB won't be 0 (default) and I'll return the amount of events for the given cache size (no automatic cache size calculation is needed).

Now... if you ONLY provided "cache-type", but no "mem-cache-size", then q.eventsCacheMemSizeMB will be 0 AND tracee calculates the amount of events for a specific cache size. Cache size depends on the amount of existing host/node memory.

Reply:
not sure we understood each other. the logics before checks 
```
if q.eventsCacheMemSizeMB > 0 {
		return AmountOfEvents(q.eventsCacheMemSizeMB)
	}
```
if it doesn't enter in the if statement will ALWAYS enter in this case case `q.eventsCacheMemSizeMB <= GBtoMB(1):` (as q.eventsCacheMemSizeMB will ALWAYS be <= 0).

so what's the point of the other switch conditions?